runtime/service.go

```diff
@@ -1261,8 +1261,13 @@ func (s *service) monitorVMExit() {
 
 	// Block until the VM exits
 	waitErr := s.machine.Wait(s.shimCtx)
-	if waitErr != nil && waitErr != context.Canceled {
-		s.logger.WithError(waitErr).Error("error returned from VM wait")
+	switch waitErr {
+	case nil:
+		s.logger.Error("the VMM has been terminated without cancelling the context")
+	case context.Canceled:
+		s.logger.Info("the VMM has been terminated by cancelling the context")
+	default:
+		s.logger.WithError(waitErr).Error("the VMM has been terminated unexpectedly")
 	}
```

Comment:
For the `context.Canceled` case, the context we're providing is the `shimCtx`, which is only going to be cancelled when the shim is shutting down, in which case the VMM is going to receive a `SIGKILL` if not already dead. So maybe a better way of approaching this would be to change `s.shimCtx` on line 1263 to just be `context.Background()` (which will never be cancelled) and then let `machine.Wait` get unblocked only by the VMM actually exiting (due to the `SIGKILL`). Then we can log whatever exit code is returned by a process exiting via `SIGKILL`.

Given that, I'm imagining something like:
```
switch waitErr {
  case nil:
     s.logger.Info("the VMM process exited with exit code 0")
  default:
   // waitErr will have the exit code or other cleanup error presumably.
   // I don't think we should say "unexpectedly" because it can happen during shim shutdown, at which time it's expected
    s.logger.WithError(waitErr).Error("the VMM process has exited")
}
```

Reply:
The way we use shimCtx is complicated. This method itself is waiting shimCtx but cancelling the context. Shutdown() is also cancelling the shim.

But, not sending SIGTERM to Firecracker may make things simpler. Let me check after adding reboot on the agent.

Reply:
Yeah the `defer s.shimCancel()` is there to handle the case where the VM process exits completely unexpectedly, but running it in other cases where the VM exited because the context was already canceled [is just a no-op](https://golang.org/pkg/context/#CancelFunc).

That's why I think it would simplify things to just use `context.Background` instead of the `shimCtx` for the call to `machine.Wait`. It should result in `Wait` only unblocking when the process has actually exited, which is easier to reason about and results in more helpful logs containing an actual exit code (`context was canceled` doesn't have much meaning to a user reading logs).

I suspect the result of having agent attempt a graceful shutdown will be that it's actually possible for `Wait()` to return a nil error (i.e. the VMM to actually exit with code 0). I think right now because we only terminate the VMM via SIGTERM [it will always exit with a non-zero code](https://unix.stackexchange.com/a/485657), meaning `Wait` should in practice never return a nil error.